### DISCUSSION TITLE
SPINNER on EVE4

### DISCUSSION THREAD

**Discussion by korstiaanS** [RudolphRiedel/FT800-FT813#22] [Q&A]
Hi,

I'm testing your library on EVE4 displays on ESP32.

Have you tried to use the spinner?
I have problems with this on EVE4. (not on EVE3 or EVE2)
Example:

This one does not work on EVE4:

```
    EVE_init_spi();
    TFT_init();

    EVE_cmd_dl(CMD_DLSTART);           // tells EVE to start a new display-list
    EVE_cmd_dl(DL_CLEAR_RGB | BLACK);  // sets the background color
    EVE_cmd_dl(DL_CLEAR | CLR_COL | CLR_STN | CLR_TAG);
    EVE_color_rgb(WHITE);
    EVE_cmd_text(5, 15, 28, 0, "Hello there!");
    EVE_cmd_spinner(200, 200, 0, 0);
    EVE_cmd_dl(DL_DISPLAY);  // put in the display list to mark its end
    EVE_cmd_dl(CMD_SWAP);    // tell EVE to use the new display list
    while (EVE_busy())
        ;
  
    while(1)
    {
        vTaskDelay(3000 / portTICK_PERIOD_MS);
    }
```

This one does work (2x spinner and a delay (200ms) needed):

```
    EVE_init_spi();
    TFT_init();

    EVE_cmd_dl(CMD_DLSTART);           // tells EVE to start a new display-list
    EVE_cmd_dl(DL_CLEAR_RGB | BLACK);  // sets the background color
    EVE_cmd_dl(DL_CLEAR | CLR_COL | CLR_STN | CLR_TAG);
    EVE_color_rgb(WHITE);
    EVE_cmd_text(5, 15, 28, 0, "Hello there!");
    EVE_cmd_spinner(200, 200, 0, 0);
    EVE_cmd_dl(DL_DISPLAY);  // put in the display list to mark its end
    EVE_cmd_dl(CMD_SWAP);    // tell EVE to use the new display list
    while (EVE_busy())
        ;
    vTaskDelay(200 / portTICK_PERIOD_MS);
    EVE_cmd_spinner(200, 200, 0, 0);

    while(1)
    {
        vTaskDelay(3000 / portTICK_PERIOD_MS);
    }
```

**Comment by korstiaanS** [comment]
Using this also works to see the spinner running (deleting the CMD_SWAP):

```
    EVE_cmd_dl(CMD_DLSTART);           // tells EVE to start a new display-list
    EVE_cmd_dl(DL_CLEAR_RGB | BLACK);  // sets the background color
    EVE_cmd_dl(DL_CLEAR | CLR_COL | CLR_STN | CLR_TAG);
    EVE_color_rgb(WHITE);
    EVE_cmd_text(5, 15, 28, 0, "Hello there!");
    EVE_cmd_spinner(EVE_HSIZE / 2, EVE_VSIZE / 2, 0, 0);
    EVE_cmd_dl(DL_DISPLAY);  // put in the display list to mark its end
    //EVE_cmd_dl(CMD_SWAP);    // tell EVE to use the new display list
    while (EVE_busy()) {
        vTaskDelay(5 / portTICK_PERIOD_MS);
    }
```

**Comment by RudolphRiedel** [comment]
Hello,

I saw the thread on the BRT forum and I have to say that I never really used CMD_SPINNER for anything.
And even after playing with it now I find it unlikely that I will use it for a project.

I just added this in tft.c of my example code:

void TFT_display(void)
{
	static int32_t rotate = 0;

 #if 1
 static uint16_t spinner_delay = 0;
 static uint8_t spinner_active = 0;
 
 spinner_delay++;
 if((spinner_delay > 100) && (spinner_delay < 200))
 {
   tft_active = 0;
   if(spinner_active == 0)
   {
     EVE_cmd_dl(CMD_DLSTART);
     EVE_cmd_dl(DL_CLEAR_RGB | BLACK);
     EVE_cmd_dl(DL_CLEAR | CLR_COL | CLR_STN | CLR_TAG);
     EVE_cmd_dl(DL_COLOR_RGB | RED);
     EVE_cmd_text(20, 20, 30, 0, "Hello there...");
     EVE_cmd_spinner(400,200,0,1);
     EVE_cmd_dl(CMD_SWAP);
     EVE_cmd_dl(DL_DISPLAY);
     spinner_active = 42;
   }    
 }
 if(spinner_delay > 200)
 {
   tft_active = 42;
   spinner_delay = 0;
   spinner_active = 0;
 }
 #endif
...

And yes, I had a CMD_SWAP in there as well and it worked with a BT815 but not with a BT817.
This way I can swap between a BT815 and a BT817 display and the display changes every 2s from
normal to spinner and back.

To be fair, even the 1.2 programming guide has this in the description of the command:
"The co-processor engine overlays the spinner on the current display list, swaps the display list to make it visible,"

So yes, they did change the behaviour a bit with BT817/BT818.

**Reply by korstiaanS** [reply]
Hi,

The latest version 2.1 of the BRT_AN033 manual under chapter 1.6 says:

> 1.6 What is new in BT817/8?
> BT817/8 maintains backward compatibility with the previous BT815/6 ICs. Any application built for BT815/6 is able to run on the BT817/8 series without any changes.
> 

They should at least mention this because all our programs (EVE2 and EVE3) with a spinner on a screen didn’t worked on EVE4 until we delete the swap command.

**Comment by RudolphRiedel** [comment]
Well, they might have not been aware that there even is an issue.
And technically, this is not strictly an issue with the BT817 but the BT815 did tolerate an
extra swap command although the documentation did say there is no need for one.

I am however wondering why we are still on the first edition of the BT817 programming guide,
it's not like there are no errors in it.
